whiterabbit/src/main/java/org/ohdsi/whiteRabbit/scan/UniformSamplingReservoir.java

```diff
@@ -0,0 +1,223 @@
+            if (removeIndex < addIndex) {
+                System.arraycopy(samples, removeIndex + 1, samples, removeIndex,
+                        addIndex - removeIndex);
+            }
+        } else if (removeIndex > addIndex) {
+            // removeIndex = 2, value = a2 -> addIndex = 1
+            // copy(1 -> 2, len 1)
+            // end: [a, a2, b, d, e]
+            System.arraycopy(samples, addIndex, samples, addIndex + 1,
+                    removeIndex - addIndex);
+        }
+
+        samples[addIndex] = value;
+    }
+
+    /**
+     * Get the quartiles of the underlying distribution. If the number of samples is larger than
+     * the maximum size of the reservoir, this will be an estimate.
+     * @return list with size three, of the 25, 50 and 75 percentiles.
+     */
+    public List<Double> getSampleQuartiles() {
+        List<Double> quartiles = new ArrayList<>(3);
+
+        switch (currentSampleLength) {
+            case 0:
+                quartiles.add(Double.NaN);
+                quartiles.add(Double.NaN);
+                quartiles.add(Double.NaN);
+                break;
+            case 1:
+                quartiles.add(samples[0]);
+                quartiles.add(samples[0]);
+                quartiles.add(samples[0]);
+                break;
+            default:
+                for (int i = 1; i <= 3; i++) {
+                    double pos = i * (currentSampleLength + 1) * 0.25d; // 25 percentile steps
+                    int intPos = (int) pos;
+                    if (intPos == 0) {
+                        quartiles.add(samples[0]);
+                    } else if (intPos == currentSampleLength) {
+                        quartiles.add(samples[currentSampleLength - 1]);
+                    } else {
+                        double diff = pos - intPos;
+                        double base = samples[intPos - 1];
+                        quartiles.add(base + diff * (samples[intPos] - base));
+                    }
+                }
+                break;
+        }
+
+        return quartiles;
+    }
+
+    public double getSampleMean() {
+        return Arrays.stream(samples).limit(currentSampleLength).sum() / currentSampleLength;
+    }
+
+    /**
+     * Get the standard deviation of the underlying distribution. If the number of samples is larger than
+     * the maximum size of the reservoir, this will be an estimate.
+     * @return double standard deviation
+     */
+    public double getSampleStandardDeviation() {
+        double sampleMean = getSampleMean();
+        double varianceSum = Arrays.stream(samples).limit(currentSampleLength).map(x -> Math.pow(x - sampleMean, 2d)).sum();
+        return Math.sqrt(varianceSum / (currentSampleLength-1));
+    }
+
+    public double getPopulationMean() {
+        if (populationCount == 0) {
+            return Double.NaN;
+        }
+        return populationSum.divide(new BigDecimal(populationCount), 8, RoundingMode.HALF_UP).doubleValue();

```

Comment:
Using a fixed scale 8 like this might mess up values with very small numbers and require unnecessary scale for very large numbers. Suggested alternative:
```
sum.scale() + count.precision()
```
The scale is the current number of digits for the sum, the `count.precision` is a proxy for the `log10` value of the `count`. This way, you add as many digits as the count would realistically justify. Example:
sum = 0.0000001 (precision 1, scale 8)
count = 1000 (precision 3, scale 0)
sum / count -> 0.0000000001 (precision 1, scale 11)

Otherwise, with a fixed scale, this would just round down to 0.

Alternatively, you can also use a.scale() + 1. The scale of the population sum will be the minimum scale of all values that actually occurred: because of the rule a+b, scale = max(a.scale, b.scale). That way the mean has as many digits as the actual values had, plus one, instead of adding digits because of the division. Adding one digit means that the mean of integers can still have a single digit.

Reply:
That is very clever, thanks! Implemented the first solution, as your alternative would indeed not work in the solution you proposed.